aten/src/ATen/core/ATenDispatch.h

```diff
@@ -103,4 +121,125 @@ class CAFFE2_API ATenDispatch {
 
 CAFFE2_API ATenDispatch& globalATenDispatch();
 
+struct lazy_holder {
+  virtual ~lazy_holder() {}
+};
+
+CAFFE2_API std::list<std::unique_ptr<lazy_holder>>& globalLazyFunctors();
+
+template <bool...>
+struct bool_pack;
+template <bool... bs>
+using all_true = std::is_same<bool_pack<bs..., true>, bool_pack<true, bs...>>;
+
+template <class R, class... Ts>
+using are_all_constructible = all_true<std::is_constructible<R, Ts>::value...>;
+
+template <typename... Ts>
+struct ivalue_constructible_tuple {
+  constexpr static bool value =
+      are_all_constructible<c10::IValue, Ts...>::value;
+};
+
+template <typename... Ts>
+struct ivalue_constructible_tuple<std::tuple<Ts...>> {
+  constexpr static bool value =
+      are_all_constructible<c10::IValue, Ts...>::value;
+};
+
+template <typename T>
+void debugIValue(std::ostream& out, T t) {
+  out << std::is_constructible<c10::IValue, T>::value << " "
+      << typeid(t).name();
+}
+
+template <typename T, typename U, typename... Args>
+void debugIValue(std::ostream& out, T t, U u, Args... args) {
+  out << std::is_constructible<c10::IValue, T>::value << " " << typeid(t).name()
+      << ',';
+  debugIValue(out, u, args...);
+}
+
+template <typename T>
+inline T& unlazy(T& t) {
+  return t;
+}
+
+template <>
+inline Tensor& unlazy<Tensor>(Tensor& t) {
+  if (c10::tensorTypeIdToBackend(t.type_id()) == Backend::Lazy) {
+    return t = t.to_eager();
+  }
+  return t;
+}
+
+template <typename R, typename... Args>
+struct lazy_wrap {
+  lazy_wrap(std::function<R(Args...)> f, const char* s) : f_(f), s_(s) {}
+  std::function<R(Args...)> f_;
+  const char* s_;
+  R operator()(Args... a) {
+    return f_(unlazy(a)...);
+  }
+};
+
+template <typename... Args>
+struct lazy_wrap<Tensor, Args...> {
+  lazy_wrap(std::function<Tensor(Args...)> f, const char* s) : f_(f), s_(s) {}
+  std::function<Tensor(Args...)> f_;
+  const char* s_;
+
+  template <
+      typename T = std::tuple<Args...>,
```

Comment:
Asking as a C++ noob: why is the `typename T = std::tuple<Args...>` needed here?

Reply:
Simpler implementation: https://godbolt.org/z/DjY4f6

Reply:
oh, just because the name of the type has to be local to the template parameters for it to be used with enable_if. In your impl the `Args...` is local to the function declaration.  If it were a class template parameter (as in this impl) you'd need my hack.  I much prefer your impl, but didn't find a great way to use it here